libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt

```diff
@@ -0,0 +1,135 @@
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.ui.lifecycle.internal.CarAppLifecycleOwner
+import com.mapbox.navigation.ui.lifecycle.internal.MapboxNavigationOwner
+
+/**
+ * Manages a default lifecycle for [MapboxNavigation].
+ *
+ * Call [MapboxNavigationApp.setup] from your application to specify your [NavigationOptions].
+ * At any point in the application you can register or unregister observers for [MapboxNavigation].
+ *
+ * Attach any and all activities to your application to share an instance of [MapboxNavigation].
+ * ```
+ * class MyApplication : Application() {
+ *   override fun onCreate() {
+ *     MapboxNavigationApp.setup(this)
+ *       .attachAllActivities()
+ *       .registerObserver(locationObserver)
+ *   }
+ * }
+ * ```
+ *
+ * Register and unregister individual LifecycleOwners. Below is an example of creating a single
+ * activity that uses [MapboxNavigation].
+ *
+ * ```
+ * class MyActivity : ComponentActivity() {
+ *   override fun onCreate() {
+ *     MapboxNavigationApp
+ *       .setup(this)
+ *       .attach(this)
+ *   }
+ *   override fun onResume() {
+ *     MapboxNavigationApp
+ *       .registerObserver(locationObserver)
+ *   }
+ *   override fun onPause() {
+ *     MapboxNavigationApp
+ *       .unregisterObserver(locationObserver)
+ *   }
+ * }
+ * ```
+ *
+ * At any point of the application you register/unregister observers
+ * [MapboxNavigationApp.registerObserver]
+ * that can observe data coming from the [MapboxNavigation] object.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+object MapboxNavigationApp {
+
+    private val carAppLifecycleOwner = CarAppLifecycleOwner()
+    private val mapboxNavigationOwner = MapboxNavigationOwner()
+
+    /**
+     * Get the lifecycle of the car and app. It is started when either the car or
+     * app is in foreground. When both the car and app have been closed, the lifecycle
+     * is stopped. The lifecycle is never destroyed.
+     */
+    val lifecycleOwner: LifecycleOwner = carAppLifecycleOwner
+
+    /**
+     * Call [MapboxNavigationApp.setup] to provide the application with [NavigationOptions].
+     */
+    fun setup(navigationOptions: NavigationOptions) = apply {
+        mapboxNavigationOwner.setup(navigationOptions)
+    }
+
+    /**
+     * Detect when any Activity is in the foreground. This can be used with [disable]
+     * if you would like to enable or disable specific user experiences.
+     *
+     * Must be called after [setup].
+     */
+    fun attachAllActivities() = apply {
+        val application = mapboxNavigationOwner.navigationOptions.applicationContext as Application
+        carAppLifecycleOwner.attachAllActivities(application)
+        carAppLifecycleOwner.lifecycle.addObserver(mapboxNavigationOwner.carAppLifecycleObserver)
+    }
```

Comment:
Using this method forces us to use `disable` manually as well, otherwise, we never get `detached` callbacks, which means that we're not taking advantage of the actual lifecycle of the components that are working with the `MapboxNavigation`, they couldn't use `onAttached`/`onDetached` to manage registration/registration with observers.

I feel like this should be the "advanced user" method and general public should use `attach`/`detach` and pass lifecycle of components that actually interact with `MapboxNavigation`.

Any thoughts on this?

Reply:
yeah have been thinking about this. I was optimizing for "mapbox navigation ON" rather than "mapbox navigation OFF"

For example, make it easy to enable everywhere - and then custom effort to disable flows. That way it has "drop in" behavior. Default to ON. But then support granular lifecycles.

You could, disable everywhere and enable a single fragment with the `MapboxNavigationApp.attach`. But then, when you add a new fragment, you need to add `MapboxNavigationApp.attach` to make it work

You could, enable everywhere with `MapboxNavigationApp.attachAllActivities`, and then disable a specific flow or event with `MapboxNavigationApp.disable()`. When you add a new fragment, it will already have `MapboxNavigation` working

But you may be right, if you are a application like AirBnB. 95% of your application will be not driving, and then "navigating to your host" will want to enable and disable a lifecycle. So I do agree with you, but what does it change?

Reply:
Right, in your big app which has a sub-feature of navigation you want the lifecycle of `MapboxNavigation` be scoped to the lifecycles that actually use it and take advantage of `MapboxNavigationObserver` to register and unregistered your listeners so that you don't leak when moving between screens and features.

More of why this currently wouldn't work is in https://github.com/mapbox/mapbox-navigation-android/pull/5112#discussion_r756784503.

Reply:
This has been updated and should be working.